doc/source/tutorial/optimize.rst

```diff
@@ -440,89 +440,225 @@ example using the Rosenbrock function follows:
-derivative) by -1 in order to perform a maximization.
+    >>> from scipy.optimize import BFGS
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac=cons_J, hess=BFGS())
 
-Then constraints are defined as a sequence of dictionaries, with keys
-``type``, ``fun`` and ``jac``.
+Alternatively, the Hessian may also be approximated using finite-differences:
 
-    >>> cons = ({'type': 'eq',
-    ...          'fun' : lambda x: np.array([x[0]**3 - x[1]]),
-    ...          'jac' : lambda x: np.array([3.0*(x[0]**2.0), -1.0])},
-    ...         {'type': 'ineq',
-    ...          'fun' : lambda x: np.array([x[1] - 1]),
-    ...          'jac' : lambda x: np.array([0.0, 1.0])})
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac=cons_J, hess='2-point')
 
+If the constraint first derivatives are also not known. They may be approximated
+by finite-difference approximation as well:
 
-Now an unconstrained optimization can be performed as:
+    >>> nonl_cons = NonlinearConstraint(cons_f, -np.inf, 1, jac='2-point', hess=BFGS())
 
-    >>> res = minimize(func, [-1.0,1.0], args=(-1.0,), jac=func_deriv,
-    ...                method='SLSQP', options={'disp': True})
-    Optimization terminated successfully.    (Exit mode 0)
-                Current function value: -2.0
-                Iterations: 4                       # may vary
-                Function evaluations: 5
-                Gradient evaluations: 4
+
+Solving the Optimization Problem:
+"""""""""""""""""""""""""""""""""
+Now with the constraints and objective functions defined, the optimization
+problem is solved using:
+
+    >>> x0 = np.array([0.5, 0])
+    >>> res = minimize(rosen, x0, method='trust-constr', jac=rosen_der, hess=rosen_hess,
+    ...                constraints=[lin_cons, nonl_cons], options={'verbose': 1}, bounds=bnds)
+    `gtol` termination condition is satisfied.
+    Number of iterations: 27, function evaluations: 17, CG iterations: 15, optimality: 8.85e-10, constraint violation: 2.22e-16, execution time: 0.044 s.
     >>> print(res.x)
-    [2. 1.]
+    [0.41494432 0.17011137]
+
+The first and second-order derivatives of the objective function may also be approximated.
+Using :func:`SR1` quasi-Newton approximation to the Hessian and finite-differences
+to approximate the gradient, the optimization problem is solved as:
+
+    >>> from scipy.optimize import SR1
+    >>> res = minimize(rosen, x0, method='trust-constr',  jac="2-point", hess=SR1(),
+    ...                constraints=[lin_cons, nonl_cons], options={'verbose': 1}, bounds=bnds)
+    `gtol` termination condition is satisfied.
+    Number of iterations: 30, function evaluations: 51, CG iterations: 16, optimality: 7.33e-09, constraint violation: 1.11e-16, execution time: 0.045 s.
+    >>> print(res.x)
+    [0.41494432 0.17011137]
+
+Finally, this method is suitable for large-scale problems because it solves the trust-region subproblem
```

Comment:
I think it would be better to create a separate section with a large-scale problem. I will try to write it immediately after the ``Solving the Optimization Problem`` section.

I will try to use a sparse Jacobian as well.

Reply:
Ok just changed my mind about that. I believe the best way would be to include some scripts on ``scipy-cookbook`` and point to there from the tutorial. I will write more about that in a comment
outside the review.